packages/loopring_v3/circuit/Gadgets/MathGadgets.h

```diff
@@ -67,15 +123,106 @@ class Constants : public GadgetT
+        pb.add_r1cs_constraint(ConstraintT(_2, FieldT::one(), FieldT(2)), ".two");
+        pb.add_r1cs_constraint(ConstraintT(_3, FieldT::one(), FieldT(3)), ".three");
+        pb.add_r1cs_constraint(ConstraintT(_4, FieldT::one(), FieldT(4)), ".four");
+        pb.add_r1cs_constraint(ConstraintT(_5, FieldT::one(), FieldT(5)), ".five");
+        pb.add_r1cs_constraint(ConstraintT(_6, FieldT::one(), FieldT(6)), ".six");
+        pb.add_r1cs_constraint(ConstraintT(_7, FieldT::one(), FieldT(7)), ".seven");
+        pb.add_r1cs_constraint(ConstraintT(_8, FieldT::one(), FieldT(8)), ".eight");
+        pb.add_r1cs_constraint(ConstraintT(_9, FieldT::one(), FieldT(9)), ".nine");
+        pb.add_r1cs_constraint(ConstraintT(_10, FieldT::one(), FieldT(10)), ".ten");
+
         pb.add_r1cs_constraint(ConstraintT(_1000, FieldT::one(), ethsnarks::FieldT(1000)), "._1000");
         pb.add_r1cs_constraint(ConstraintT(_1001, FieldT::one(), ethsnarks::FieldT(1001)), "._1001");
         pb.add_r1cs_constraint(ConstraintT(_10000, FieldT::one(), ethsnarks::FieldT(10000)), "._10000");
         pb.add_r1cs_constraint(ConstraintT(_100000, FieldT::one(), ethsnarks::FieldT(100000)), "._100000");
+        pb.add_r1cs_constraint(ConstraintT(indexBase, FieldT::one(), ethsnarks::FieldT(INDEX_BASE)), ".indexBase");
         pb.add_r1cs_constraint(ConstraintT(emptyTradeHistory, FieldT::one(), ethsnarks::FieldT(EMPTY_TRADE_HISTORY)), ".emptyTradeHistory");
         pb.add_r1cs_constraint(ConstraintT(maxAmount, FieldT::one(), ethsnarks::FieldT(MAX_AMOUNT)), ".maxAmount");
         pb.add_r1cs_constraint(ConstraintT(maxConcurrentOrderIDs, FieldT::one(), ethsnarks::FieldT(MAX_CONCURRENT_ORDERIDS)), ".maxConcurrentOrderIDs");
+        pb.add_r1cs_constraint(ConstraintT(dummyPublicKeyX, FieldT::one(), ethsnarks::FieldT("132404916167441185773716937639098950030214269269071041759116060313694190797")), ".dummyPublicKeyX");
+        pb.add_r1cs_constraint(ConstraintT(dummyPublicKeyY, FieldT::one(), ethsnarks::FieldT("6933274320914065805670637410453081675154127044926882796951068647148079547843")), ".dummyPublicKeyY");
+        pb.add_r1cs_constraint(ConstraintT(txTypeTransfer, FieldT::one(), ethsnarks::FieldT(int(TransactionType::Transfer))), ".txTypeTransfer");
+        pb.add_r1cs_constraint(ConstraintT(txTypeNewAccount, FieldT::one(), ethsnarks::FieldT(int(TransactionType::AccountNew))), ".txTypeNewAccount");
+
+        pb.add_r1cs_constraint(ConstraintT(pow10_c0, FieldT::one(), ethsnarks::FieldT(POW10_C0)), ".pow10_c0");
+        pb.add_r1cs_constraint(ConstraintT(pow10_c1, FieldT::one(), ethsnarks::FieldT(POW10_C1)), ".pow10_c1");
+        pb.add_r1cs_constraint(ConstraintT(pow10_c2, FieldT::one(), ethsnarks::FieldT(POW10_C2)), ".pow10_c2");
+        pb.add_r1cs_constraint(ConstraintT(pow10_c3, FieldT::one(), ethsnarks::FieldT(POW10_C3)), ".pow10_c3");
+    }
+};
+
+class DualVariableGadget : public libsnark::dual_variable_gadget<FieldT>
```

Comment:
Sure, but doesn't really matter for the performance, especially for these extremely small vectors (probably implementation dependent, but I remember vectors being optimized for small lengths so probably the memory is already there either way).

Creating circuits is also a pretty slow operation because it's a sequential process, which is a problem the server mode solves by only creating the circuits once and reusing them for all subsequent proof generations, so in the best case this code isn't run often and is not important for the prover times.